crates/lind-common/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+use std::sync::{Arc, Barrier, Condvar, Mutex};
+use wasmtime::{AsContext, AsContextMut, Caller, ExternType, Linker, Module, SharedMemory, Store, Val, Extern, OnCalledAction, RewindingReturn, StoreOpaque, InstanceId};
+
+// lind-common serves as the main entry point when lind_syscall. Any syscalls made in glibc would reach here first,
+// then the syscall would be dispatched into rawposix, or other crates under wasmtime, depending on the syscall, to perform its job
+
+#[derive(Clone)]
+// stores some attributes associated with current runnning wasm instance (i.e. cage)
+// each cage has its own lind-common context 
+pub struct LindCommonCtx {
+    // process id attached to the lind-common context, should be same as cage id
+    pid: i32,
+    
+    // next cage id, shared between all lind-common context instance (i.e. all cages)
+    next_cageid: Arc<AtomicU64>,
+}
+
+impl LindCommonCtx {
+    // create a new lind-common context, should only be called once for then entire runtime
+    pub fn new(next_cageid: Arc<AtomicU64>) -> Result<Self> {
+        // cage id starts from 1
+        let pid = 1;
+        Ok(Self { pid, next_cageid })
+    }
+
+    // create a new lind-common context with pid provided, used by exec syscall
+    pub fn new_with_pid(pid: i32, next_cageid: Arc<AtomicU64>,
+        ) -> Result<Self> {
+        Ok(Self { pid, next_cageid })
+    }
+
+    // entry point for lind_syscall in glibc, dispatching syscalls to rawposix or wasmtime
+    pub fn lind_syscall<T: LindHost<T, U> + Clone + Send + 'static + std::marker::Sync, U: Clone + Send + 'static + std::marker::Sync>
+                        (&self, call_number: u32, call_name: u64, caller: &mut Caller<'_, T>, arg1: u64, arg2: u64, arg3: u64, arg4: u64, arg5: u64, arg6: u64) -> i32 {
+        let start_address = get_memory_base(&caller);
+        match call_number as i32 {
+            // clone syscall
+            171 => {
+                let clone_args = unsafe { &mut *((arg1 + start_address) as *mut CloneArgStruct) };
+                clone_args.child_tid += start_address;
+                wasmtime_lind_multi_process::clone_syscall(caller, clone_args)
+            }
+            // exec syscall
+            69 => {
+                wasmtime_lind_multi_process::exec_syscall(caller, arg1 as i64, arg2 as i64, arg3 as i64)
+            }
+            // exit syscall
+            30 => {
+                wasmtime_lind_multi_process::exit_syscall(caller, arg1 as i32)
+            }
+            // other syscalls goes into rawposix
+            _ => {
+                lind_syscall_api(
+                    self.pid as u64,
+                    call_number,
+                    call_name,
+                    start_address,
+                    arg1,
+                    arg2,
+                    arg3,
+                    arg4,
+                    arg5,
+                    arg6,
+                )
+            }
+        }
+    }
+
+    // get current process id/cageid
```

Comment:
does this not call into rawposix? if not what is it for?

Reply:
This function is currently not used. I just created an interface to get the pid for the context, in case I need it in the future

Reply:
ok, maybe note that. it may get a bit confusing if were storing pid here and in rawposix separately, though they both should be immutable.